cli/device.cpp

```diff
@@ -368,20 +379,41 @@ bool VulkanDevice::init_device(const Options &opts)
 	const auto *requested_pdf2 = opts.features;
 	VkPhysicalDeviceFeatures2 replacement_pdf2;
 	VkPhysicalDeviceRobustness2FeaturesEXT replacement_robustness2;
+	VkPhysicalDeviceFragmentShadingRateFeaturesKHR replacement_fragment_shading_rate;
 
-	if (opts.features && application_info_promote_robustness2(opts.application_info) &&
-	    find_pnext<VkPhysicalDeviceRobustness2FeaturesEXT>(
-			    VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT,
-			    opts.features->pNext) == nullptr)
+	if (opts.features)
 	{
-		replacement_robustness2 = { VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT };
-		replacement_pdf2 = *opts.features;
-		replacement_pdf2.pNext = &replacement_robustness2;
-		replacement_robustness2.pNext = opts.features->pNext;
-		replacement_robustness2.robustBufferAccess2 = opts.features->features.robustBufferAccess;
-		replacement_robustness2.robustImageAccess2 = opts.features->features.robustBufferAccess;
-		replacement_robustness2.nullDescriptor = VK_TRUE;
-		requested_pdf2 = &replacement_pdf2;
+		if (application_info_promote_robustness2(opts.application_info) &&
+		    find_pnext<VkPhysicalDeviceRobustness2FeaturesEXT>(
+				    VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT,
+				    opts.features->pNext) == nullptr)
+		{
+			replacement_robustness2 = { VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_ROBUSTNESS_2_FEATURES_EXT };
+			replacement_pdf2 = *opts.features;
+			replacement_pdf2.pNext = &replacement_robustness2;
+			replacement_robustness2.pNext = opts.features->pNext;
+			replacement_robustness2.robustBufferAccess2 = opts.features->features.robustBufferAccess;
+			replacement_robustness2.robustImageAccess2 = opts.features->features.robustBufferAccess;
+			replacement_robustness2.nullDescriptor = VK_TRUE;
+
+			requested_pdf2 = &replacement_pdf2;
+		}
+
+		if (application_info_promote_fragment_shading_rate(opts.application_info) &&
+		    find_pnext<VkPhysicalDeviceFragmentShadingRateFeaturesKHR>(
+				    VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADING_RATE_FEATURES_KHR,
+				    opts.features->pNext) == nullptr)
+		{
+			replacement_fragment_shading_rate = { VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADING_RATE_FEATURES_KHR };
+			replacement_pdf2 = *requested_pdf2;
```

Comment:
The branch above copies PDF2 data from opts.features. Probably a good idea to harmonize the two.

Reply:
Sounds good.